File: sdk/src/main/java/com/cxense/cxensesdk/model/ExternalItem.java
package com.cxense.cxensesdk.model;

import androidx.annotation.NonNull;

import com.cxense.cxensesdk.Preconditions;
import com.google.gson.annotations.SerializedName;

import java.util.Locale;

/**
 * Stored key-values for the user.
 *
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-06-05).
 */
@SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
public final class ExternalItem {
    private static final String GROUP_REGEXP = "^[a-zA-Z\\d-]{1,%d}$";
    private static final int GROUP_NAME_MAX_LENGTH = 26;
    private static final int ITEM_NAME_MAX_LENGTH = 100;
    private static final String ERROR_MSG = "'%s' can contains only letters, digits or dash, max length is %d";
    @SerializedName("group")
    String group;
    @SerializedName("item")
    String item;

    private ExternalItem() {
    }

    ExternalItem(@NonNull String group, @NonNull String item) {
        this();
        Preconditions.checkStringForRegex(group, "group", String.format(Locale.US, GROUP_REGEXP,
                GROUP_NAME_MAX_LENGTH), ERROR_MSG, "group", GROUP_NAME_MAX_LENGTH);
        Preconditions.checkStringMaxLength(item, "item", ITEM_NAME_MAX_LENGTH);
        this.group = group;
        this.item = item;
    }

    /**
     * Gets the group name.
     *
     * @return group name
     */
    @NonNull
    public String getGroup() {
        return group;
    }

    /**
     * Gets the item which is to be associated with the group name.
     *
     * @return item
     */
    @NonNull
    public String getItem() {
        return item;
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/model/ExternalItem.kt
package com.cxense.cxensesdk.model

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * Stored key-values for the user.
 * @param group group name
 * @property group group name
 *
 * @param item the item which is to be associated with the group name.
 * @property item the item which is to be associated with the group name.
 *
 * @throws IllegalArgumentException if [group] or [item] doesn't meet a criteria.
 */
@Suppress("unused", "MemberVisibilityCanBePrivate") // Public API.
@JsonClass(generateAdapter = true)
class ExternalItem(
    group: String,
    item: String
) {
    @Json(name = "group")
    val group: String = group.also {
        require(it.matches(GROUP_REGEXP.toRegex())) {
            "Group should not be empty and can contains only letters, digits or dash," +
                " max length is $GROUP_NAME_MAX_LENGTH"
        }
    }

    @Json(name = "item")
    val item: String = item.also {
        require(it.length <= ITEM_NAME_MAX_LENGTH) {
            "Item can't be longer than $ITEM_NAME_MAX_LENGTH symbols"
        }
    }

    companion object {
        private const val GROUP_NAME_MAX_LENGTH = 26
        private const val GROUP_REGEXP = "^[a-zA-Z\\d-]{1,$GROUP_NAME_MAX_LENGTH}$"
        private const val ITEM_NAME_MAX_LENGTH = 100
    }
}
